### DISCUSSION TITLE
Tableau HTML/CSS

### DISCUSSION THREAD

**Discussion by JeremieDENAEGEL** [witsa/synapps#745] [Q&A]
Bonjour,

J'ai un tableau HTML/CSS qui affiche des données de compteurs.
Actuellement j'ai un additionnel pour chaque case du tableau.

Est-il possible de faire les liaisons directement dans le contenu HTML de l'acteur sans passer par la case additionnels ? 
En sachant que les données sont liées à des fournisseurs relatifs de variable différents pour chaque ligne

![Tableau01](https://github.com/witsa/synapps/assets/50445069/f43a20d4-6ca6-4b10-9e53-f15549717f32)

Merci d'avance

**Comment by manufitoussiwit** [comment]
Salut, j'essais des solutions et je reviens vers toi dès qu'il y en a une satisfaisante ;-)

**Comment by manufitoussiwit** [comment]
Salut, 

Alors j'ai une proposition. J'en avais plusieurs à coup de javascript. Voila celle qui en utilise pratiquement pas.

L'idée est d'utiliser un composite pour aller chercher les données qui se répètent sur chaque ligne. J'ai vu que l'entrée de chaque ligne, c'est une ressource. Ensuite les données des colonnes sont prises sur des ressources/fonctions en dessus de la ressource.

Alors j'ai fais le composite suivant : 

![image](https://github.com/user-attachments/assets/cc5d8614-1dc3-4d7e-9924-3735eaf02e80)

En propriété du composite, on a : 

![image](https://github.com/user-attachments/assets/60a029d7-2296-4fab-aa14-1731614aa588)

- `path` est le chemin depuis :easy.RESS de la ressource qui porte les données de la ligne du tableau.
- `row` qui est en lecture seule va permettre de récupérer par liaison la ligne html de tableau avec les données bien renseignées et synchro en plus.

## Pour comprendre comment c'est généré

Dans le composite, on fait autant d'acteur variable relatif que de colonne dans le tableau.
![image](https://github.com/user-attachments/assets/68ea6acd-5f9a-4da7-8368-81c8f9a5d377)

`data1` par exemple va chercher le nom de la ressource (1x d'ailleurs)

![image](https://github.com/user-attachments/assets/51d91a94-b952-4cc2-8bf1-21173523154f)

`data2` va chercher la donnée de la première colonne (à chaque rafraichissement): 

![image](https://github.com/user-attachments/assets/ebf97a72-9c5f-4047-a5f4-63f3ba94e657)

On fait la même chose pour chaque colonne.

Ensuite, l'acteur `html4` va permettre plusieurs chose :

![image](https://github.com/user-attachments/assets/fa1aa53a-b971-4434-aeb0-56ab0eece332)

![image](https://github.com/user-attachments/assets/9d8d6ec8-8d46-4626-a9e3-336bef8b03bd)


Il va permettre de construire la ligne html de tableau : 

![image](https://github.com/user-attachments/assets/daf815c4-04df-45ef-ad9e-50b477342c12)

avec des jokers, un par valeur de colonne. Au passage, on notera que j'ai ajouté une classe CSS. C'est pour personnaliser la cellule plus tard.

Ensuite, on voit qu'il y a une liaison vers les données que je récupère avec les acteurs relatifs.

Pour finir, on voit une additionnelle qui sert à exposer la ligne construite sur la proprité `row` du composite.

## Maintenant dans la scène : 

![image](https://github.com/user-attachments/assets/fcbb2b55-8789-40e0-8a23-21e16252a935)

On ajoute autant de composite que de ligne à ajouter.

Chaque composite va aller chercher la ligne pour une ressource : 

![image](https://github.com/user-attachments/assets/b3e63d77-55d3-4192-b78b-e5fa79b9d934)

Ensuite, l'acteur html `tab` va permettre de construire le tableau : 

![image](https://github.com/user-attachments/assets/32ed654a-c62b-4ace-91b5-42c41b87b002)

Le contenu : 

![image](https://github.com/user-attachments/assets/2ac5fa05-1323-4474-ad1d-ea278a1af70c)

Il y a un joker par ligne à placer dans le tableau. Au passage on voit le CSS pour personnaliser le tableau et les cellules de nombre.

Chaque joker est finalement lié à son composite.

Et voila ! 

![image](https://github.com/user-attachments/assets/21231212-c83d-4ba8-95ad-06fbb1cf6ac9)

Je pense que tu devrais pouvoir adapter ça à ton cas ;-) 

## La scène 
```text
SYNAPPS-STUDIO-SCENE|{"config":{"key":"scene124","name":"Exemple de tableau de données"},"leadActor":{"type":"layout/stack","key":"stack1","children":[{"type":"composite/data-row","key":"data-row1"},{"type":"composite/data-row","key":"data-row2","properties":{"path":"R00039.R00041"}},{"type":"composite/data-row","key":"data-row3","properties":{"path":"R00039.R00042"}},{"type":"composite/data-row","key":"data-row4","properties":{"path":"R00039.R00043"}},{"type":"display/html","key":"tab","properties":{"content":"<table class=\"data\">\n  <thead>\n    <tr>\n      <th>Nom des compteurs</th>\n      <th>Conso. d'aujourd'hui</th>\n      <th>Conso. d'hier</th>\n      <th>Conso. mois en cours</th>\n    </tr>\n  </thead>\n  <tbody>\n    {{row1}}\n    {{row2}}\n    {{row3}}\n    {{row4}}\n  </tbody>\n</table>\n\n<style>\n  table.data tr th {\n    font-weight: bold;\n    color: #6699cc;\n  }\n\n  table.data tr td,\n  table tr th {\n    border: 1px solid #aaa;\n    padding: 5px 10px;\n  }\n\n  table.data tbody tr:nth-child(odd) {\n    background-color: #99bbee;\n  }\n\n  table.data tbody tr:nth-child(even) {\n    background-color: #6699cc;\n  }\n\n  table.data tbody tr td.col2,\n  table.data tbody tr td.col3,\n  table.data tbody tr td.col4 {\n    text-align: right;\n  }\n</style>","horizontalAlignment":"middle","verticalAlignment":"middle"},"additionalDefs":{"row1":{"type":"text"},"row2":{"type":"text"},"row3":{"type":"text"},"row4":{"type":"text"}},"bindings":{"additionals.row1":"actor#data-row1@properties.row","additionals.row2":"actor#data-row2@properties.row","additionals.row3":"actor#data-row3@properties.row","additionals.row4":"actor#data-row4@properties.row"}}]}}
```

## Le composite 
```text
SYNAPPS-STUDIO-COMPOSITE|{"config":{"key":"data-row","name":"Ligne de données","additionalDefs":{"path":{"type":"wos-path","value":"R00039.R00040","relativeToPath":":easy.RESS","variableType":"resourcesOnly"},"row":{"type":"text","isReadOnly":true,"value":""}}},"leadActor":{"type":"layout/stack","key":"root","children":[{"type":"redy/data-source/wos-relative-variable","key":"resource","properties":{"dataReadMode":"always","relativeTo":"global/ress","relativePath":"R00039.R00040"},"bindings":{"properties.relativePath":"stage@properties.path"}},{"type":"redy/data-source/wos-relative-variable","key":"data1","properties":{"fieldName":"name","relativeTo":"resource"}},{"type":"redy/data-source/wos-relative-variable","key":"data2","properties":{"fieldName":"value","relativeTo":"resource","relativePath":"FNCT.F0001.In","dataReadMode":"always"}},{"type":"redy/data-source/wos-relative-variable","key":"data3","properties":{"fieldName":"value","relativeTo":"resource","relativePath":"FNCT.F0002.In","dataReadMode":"always"}},{"type":"redy/data-source/wos-relative-variable","key":"data4","properties":{"fieldName":"value","relativeTo":"resource","relativePath":"FNCT.F0003.In","dataReadMode":"always"}},{"type":"display/html","key":"html1","properties":{"content":"<tr>\n  <td class=\"col1\">{{data1}}</td>\n  <td class=\"col2\">{{data2}}</td>\n  <td class=\"col3\">{{data3}}</td>\n  <td class=\"col4\">{{data4}}</td>\n</tr>","visible":false},"additionalDefs":{"data1":{"type":"text"},"data2":{"type":"text"},"data3":{"type":"text"},"data4":{"type":"text"},"completedContent":{"type":"text"}},"bindings":{"additionals.data1":"actor#data1@data","additionals.data2":"actor#data2@data","additionals.data3":"actor#data3@data","additionals.data4":"actor#data4@data","additionals.completedContent":{"canWrite":true,"scripts":[{"path":"additionals/completedContent/binding/onWriteTransform","value":"return this.completedContent;"}],"source":"stage@properties.row"}},"events":{"additionals/completedContent/binding/onWriteTransform":["return this.completedContent;"],"onRender":["this.additionals.completedContent = this.completedContent;"]}}]}}
```

**Comment by sebusone** [comment]
Bonjour à tous, 

J'aimerais savoir comment tu as procédé pour obtenir ces valeurs (conso mois, mois -1 etc.). de coté de l'automate? Ceci est un script ou une ressource disponible de base? 

Merci pour ton aide.

**Comment by JeremieDENAEGEL** [comment]
Oula, j'avais opté pour la version brute.
Une avalanche de bloc fonction dans une ressource (de mémoire, j'en avais au moins 150 😂)

Je ferai peut-être différement maintenant 😂

**Reply by sebusone** [reply]
Tu aurais une idée, ou quelqu'un de forum, comment pourrait-t-on faire pour obtenir ce résultat? 
Il existe peut-être une solution déjà étudié par les équipes?

**Reply by JeremieDENAEGEL** [reply]
Que cherches-tu à faire et quel est ton besoin ?

**Reply by sebusone** [reply]
J'aimerais obtenir à peu près le même tableau que toi. Côte Synapps pas de difficulté, cependant je ne sais pas du tout comment le jouer de côte de la prog, pour la mémorisation des données, pour le déclanchement de mémorisation etc.

**Comment by JeremieDENAEGEL** [comment]
La mémorisation des données est la partie la plus difficile surtout si le client utilise les données pour de la facturation.

J'ai opté pour le bon vieux bloc fonction Delta dans mon cas.

Voici la version la plus légère que j'ai fait pour mon tableau côté prog 😂.
Normalement il y a 2 à 3 fois plus de blocs fonctions 😂.

Il y a probablement d'autres solutions

![Capture d’écran 2025-05-02 101016](https://github.com/user-attachments/assets/8b9af98d-e921-4953-82a4-d64a879598e8)

**Reply by sebusone** [reply]
Je te remercie pour ton partage. 
Ca donne des idées effectivement.
